test/clj/rems/test_browser.clj

```diff
@@ -153,157 +253,166 @@
-        ;; leave "Text field with max length" empty
-        ;; leave "Text are with max length" empty
-
-        (accept-licenses)
-        (send-application)
-        (is (= "Applied" (get-element-text *driver* :application-state)))
-
-        (testing "check a field answer"
-          (is (= "Test name" (get-element-text *driver* description-field-selector))))
-
-        (testing "see application on applications page"
-          (go-to-applications)
-          (let [summary (get-application-summary application-id)]
-            (is (= "Default workflow" (:resource summary)))
-            (is (= "Applied" (:state summary)))
-            ;; don't bother trying to predict the external id:
-            (is (.contains (:description summary) "Test name"))))
-
-        (testing "fetch application from API"
-          (let [application (:body
-                             (http/get (str +test-url+ "/api/applications/" application-id)
-                                       {:as :json
-                                        :headers {"x-rems-api-key" "42"
-                                                  "x-rems-user-id" "handler"}}))
-                attachment-id (get-in application [:application/attachments 0 :attachment/id])]
-            (testing "attachments"
-              (is (= [{:attachment/id attachment-id
-                       :attachment/filename "test.txt"
-                       :attachment/type "text/plain"}]
-                     (:application/attachments application))))
-            (testing "applicant information"
-              (is (= "alice" (get-in application [:application/applicant :userid])))
-              (is (= (set (map :license/id (:application/licenses application)))
-                     (set (get-in application [:application/accepted-licenses :alice])))))
-            (testing "form fields"
-              (is (= "Test name" (:application/description application)))
-              (is (= [["label" ""]
-                      ["description" "Test name"]
-                      ["text" "Test"]
-                      ["texta" "Test2"]
-                      ["header" ""]
-                      ["date" "2050-01-02"]
-                      ["email" "user@example.com"]
-                      ["attachment" (str attachment-id)]
-                      ["option" "Option1"]
-                      ["text" "Conditional"]
-                      ["multiselect" "Option2 Option3"]
-                      ["label" ""]
-                      ["text" ""]
-                      ["texta" ""]]
-                     (for [field (select [:application/forms ALL :form/fields ALL] application)]
-                       ;; TODO could test other fields here too, e.g. title
-                       [(:field/type field)
-                        (:field/value field)]))))
-            (testing "after navigating to the application view again"
-              (scroll-and-click *driver* [{:css "table.my-applications"}
+  (let [driver (get-driver)]
```

Comment:
I can see the need for consistently passing in the driver as the first argument for everything
... but I'm not sure if `(let [driver (get-driver)] (doto driver ...))` is any better than `(doto *driver* ...)`

Did you consider wrapping all etaoin functions so that they get a `*driver*` parameter? This is what we do with `*db*` and it works nicely enough.

Not insisting on any changes, just thinking out loud.

Reply:
Like discussed in the meeting, I will dump the utilities into their own namespace and then refactor it so that we can use a dynamic driver again and don't need the `doto` or `let` `get-driver` anymore. It will be much better.